### PR TITLE
Full Text Search using Solr Jetty

### PR DESCRIPTION
**Pull request by jvolkenant**
PR adds full text search using solr-jetty package instead of tomcat like #1506.  I've been using this PR for a few months and it's been working well.  Search in Webmail, Android IMAP (I use Nine on Android), and with Z-Push 2.5.0, Nine on Android works when using Exchange.

I feel like solr-jetty uses less memory than the Tomcat version does.

```
root@m:~# find /home/user-data/mail/mailboxes/ | wc -l
29063
root@m:~# uptime
 10:39:49 up 2 days,  3:28,  1 user,  load average: 0.27, 0.13, 0.10
root@m:~# free -m
              total        used        free      shared  buff/cache   available
Mem:            985         564         115          59         305         218
Swap:          1535          56        1479
```

### CODE REVIEW

File: setup/solr.sh [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+# PERMISSIONS
+
+# Ensure configuration files are owned by dovecot and not world readable.
+chown -R mail:dovecot /etc/dovecot
+chmod -R o-rwx /etc/dovecot
+
+# Newer updates to jetty9 restrict write directories, this allows for
+# jetty to write to solr database directories
+cat > /etc/systemd/system/jetty9.service.d/solr-permissions.conf << EOF
```
Comment (yodax) [inline setup/solr.sh]: My box required:

```suggestion
mkdir -p /etc/systemd/system/jetty9.service.d/
cat > /etc/systemd/system/jetty9.service.d/solr-permissions.conf << EOF
```

**Comment by yodax** [comment]
load-day
cpu-day
memory-day
interrupts-day
Since installing this this morning you can see a bump in resource usage. My guess is that this could be an issue for the lowerend hardware.

I think this is worth it, but I thought I would post it here.

**Comment by yodax** [comment]
The box has settled down. I think this spike might have been due to the new z-push install that was syncing the client I was testing with.

File: setup/solr.sh (hunk 1 above)
Comment (jvolkenant) [reply setup/solr.sh]: Thanks for testing, I've added your recommended change.

**Comment by yodax** [comment]
I've been running this on my main box for a while now. Works great.

**Comment by jvolkenant** [comment]
I will admit that on my box which has 985MB of ram, Jetty is using 297MB itself. I'm not sure if it's 30% of my memory based on how much email's I have indexed, a percentage of memory or a set amount.  In either case.  We have reduced the minimum amount of memory that we will allow MIAB to run on.  Adding this to the default install might hurt more than help.

Perhaps adding a memory check to solr.sh to only install if $TOTAL_PHYSICAL_MEM -gt X?

**Comment by yodax** [comment]
I was digging through the mail log and I saw these: 

```Jul 17 07:36:04 indexer-worker(x@x.email): Error: fts_solr: Indexing failed: 404 Not Found
Jul 17 07:36:04 indexer-worker(x@x.email): Error: Mailbox Archive: Transaction commit failed: FTS transaction commit failed: backend deinit (attempted to index 1001 messages (UIDs 9..2958))```

jetty-console.log has nothing for that time. It does have one config error:

```Jul 17 01:21:32 box jetty9[1056]: Running Jetty:
Jul 17 01:21:36 box jetty9[1056]: 2019-07-17 01:21:36.428:INFO::main: Logging initialized @4008ms to org.eclipse.jetty.util.log.StdErrLog
Jul 17 01:21:36 box jetty9[1056]: 2019-07-17 01:21:36.758:WARN:oejx.XmlConfiguration:main: Property 'threads.min' is deprecated, use 'jetty.threadPool.minThreads' instead
Jul 17 01:21:36 box jetty9[1056]: 2019-07-17 01:21:36.765:WARN:oejx.XmlConfiguration:main: Property 'threads.max' is deprecated, use 'jetty.threadPool.maxThreads' instead
Jul 17 01:21:36 box jetty9[1056]: 2019-07-17 01:21:36.775:WARN:oejx.XmlConfiguration:main: Property 'threads.timeout' is deprecated, use 'jetty.threadPool.idleTimeout' instead
Jul 17 01:21:37 box jetty9[1056]: 2019-07-17 01:21:37.172:WARN:oejx.XmlConfiguration:main: Property 'jetty.port' is deprecated, use 'jetty.http.port' instead
Jul 17 01:21:37 box jetty9[1056]: 2019-07-17 01:21:37.173:WARN:oejx.XmlConfiguration:main: Property 'http.timeout' is deprecated, use 'jetty.http.idleTimeout' instead
Jul 17 01:21:37 box jetty9[1056]: 2019-07-17 01:21:37.244:INFO:oejs.Server:main: jetty-9.4.15.v20190215; built: unknown; git: unknown; jvm 11.0.3+7-Ubuntu-1ubuntu218.04.1
Jul 17 01:21:37 box jetty9[1056]: 2019-07-17 01:21:37.324:INFO:oejdp.ScanningAppProvider:main: Deployment monitor [file:///var/lib/jetty9/webapps/] at interval 1
Jul 17 01:21:37 box jetty9[1056]: 2019-07-17 01:21:37.904:INFO:oeja.AnnotationConfiguration:main: Scanning elapsed time=1ms
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.026:INFO:oejshC.root:main: Warning: No org.apache.tomcat.JarScanner set in ServletContext. Falling back to default JarScanner implementation.
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.308:INFO:oajs.TldScanner:main: At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.546:INFO:oejs.session:main: DefaultSessionIdManager workerName=node0
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.547:INFO:oejs.session:main: No SessionScavenger set, using defaults
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.549:INFO:oejs.session:main: node0 Scavenging every 660000ms
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.632:INFO:oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@2b662a77{root,/,file:///var/lib/jetty9/webapps/root/,AVAILABLE}{/root}
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.647:WARN:oejx.XmlConfiguration:main: Config error at <Call name="addAliasCheck"><Arg>|      <New class="org.eclipse.jetty.server.handler.ContextHandler$ApproveSameSuffixAliases"/>|    </Arg></Call> java.lang.ClassNotFoundException: org.eclipse.jetty.server.handler.ContextHandler$ApproveSameSuffixAliases in file:/etc/solr/solr-jetty.xml
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.647:WARN:oejd.DeploymentManager:main: Unable to reach node goal: started
Jul 17 01:21:38 box jetty9[1056]: java.lang.ClassNotFoundException: org.eclipse.jetty.server.handler.ContextHandler$ApproveSameSuffixAliases
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.Loader.loadClass(Loader.java:65)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.newObj(XmlConfiguration.java:834)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.itemValue(XmlConfiguration.java:1309)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.value(XmlConfiguration.java:1214)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.access$500(XmlConfiguration.java:326)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$AttrOrElementNode.getList(XmlConfiguration.java:1442)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$AttrOrElementNode.getList(XmlConfiguration.java:1417)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:780)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:472)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:413)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:311)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.providers.WebAppProvider.createContextHandler(WebAppProvider.java:301)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.App.getContextHandler(App.java:105)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.bindings.StandardDeployer.processBinding(StandardDeployer.java:38)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.AppLifeCycle.runBindings(AppLifeCycle.java:192)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.DeploymentManager.requestAppGoal(DeploymentManager.java:505)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.DeploymentManager.addApp(DeploymentManager.java:151)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.providers.ScanningAppProvider.fileAdded(ScanningAppProvider.java:180)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.providers.WebAppProvider.fileAdded(WebAppProvider.java:453)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.providers.ScanningAppProvider$1.fileAdded(ScanningAppProvider.java:64)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.Scanner.reportAddition(Scanner.java:610)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.Scanner.reportDifferences(Scanner.java:529)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.Scanner.scan(Scanner.java:392)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.Scanner.doStart(Scanner.java:313)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.providers.ScanningAppProvider.doStart(ScanningAppProvider.java:150)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.DeploymentManager.startAppProvider(DeploymentManager.java:579)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.deploy.DeploymentManager.doStart(DeploymentManager.java:240)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:138)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.server.Server.start(Server.java:415)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:113)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.server.Server.doStart(Server.java:382)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$1.run(XmlConfiguration.java:1572)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration$1.run(XmlConfiguration.java:1512)
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/java.security.AccessController.doPrivileged(Native Method)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1511)
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Jul 17 01:21:38 box jetty9[1056]: #011at java.base/java.lang.reflect.Method.invoke(Method.java:566)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.start.Main.invokeMain(Main.java:220)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.start.Main.start(Main.java:490)
Jul 17 01:21:38 box jetty9[1056]: #011at org.eclipse.jetty.start.Main.main(Main.java:77)
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.712:INFO:oejs.AbstractConnector:main: Started ServerConnector@c94fd30{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.713:INFO:oejs.Server:main: Started @6293ms```

Running the cron manually results in:
```/usr/bin/curl http://127.0.0.1:8080/solr/update?commit=true
<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
<title>Error 404 Not Found</title>
</head>
<body><h2>HTTP ERROR 404</h2>
<p>Problem accessing /solr/update. Reason:
<pre>    Not Found</pre></p><hr><a href="http://eclipse.org/jetty">Powered by Jetty:// 9.4.15.v20190215</a><hr/>

</body>
</html>```

It seems that it's not working. I had the feeling that the search was faster, maybe it was the hardware upgrade.

I checked the script, I don't see where we hookup solr to jetty, is that handled by the package?

**Comment by jvolkenant** [comment]
Did you get this on an existing install that had solr.sh used previously or a new install?

This error is curious, does this file match mine below?  It is commented out in my files from the installed package. 
```
Jul 17 01:21:38 box jetty9[1056]: 2019-07-17 01:21:38.647:WARN:oejx.XmlConfiguration:main: Config error at <Call name="addAliasCheck"><Arg>|      <New class="org.eclipse.jetty.server.handler.ContextHandler$ApproveSameSuffixAliases"/>|    </Arg></Call> java.lang.ClassNotFoundException: org.eclipse.jetty.server.handler.ContextHandler$ApproveSameSuffixAliases in file:/etc/solr/solr-jetty.xml
```

My /etc/solr/solr-jetty.xml looks like this (unmodified from solr-jetty install)
```
<?xml version="1.0"  encoding="ISO-8859-1"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure.dtd">

<!-- Context configuration file for the Solr web application in Jetty -->

<Configure class="org.eclipse.jetty.webapp.WebAppContext">
  <Set name="contextPath">/solr</Set>
  <Set name="war">/usr/share/solr/web</Set>

  <!-- Set the solr.solr.home system property -->
  <Call name="setProperty" class="java.lang.System">
    <Arg type="String">solr.solr.home</Arg>
    <Arg type="String">/usr/share/solr</Arg>
  </Call>

  <!-- Enable symlinks -->
  <!--  <Call name="addAliasCheck">
    <Arg>
      <New class="org.eclipse.jetty.server.handler.ContextHandler$ApproveSameSuffixAliases"/>
    </Arg>
  </Call>-->
</Configure>
```

I believe that jetty gets the directive for /solr from one of these two paths (files look to have the same contents for me)
```
root@m:/var/log# dpkg-query -L solr-jetty | grep xml 
/etc/solr/solr-jetty.xml
/var/lib/jetty9/webapps/solr.xml
```

I am not that familiar with java/jetty, I could only recommend apt-get purge solr-jetty and rerunning the install.

**Comment by yodax** [comment]
Thanks for the assist.

>Did you get this on an existing install that had solr.sh used previously or a new install?
This is a long running box, the change went on there clean. 

The files are identical.

I tried running the purge and reinstall, this didn't work. I haven't tweaked anything regarding solr-jetty. 

I'll have a look and see if I can find the root cause.

**Comment by yodax** [comment]
I tried figuring out why this didn't work for me. Sadly I couldn't get to the root cause. I have reverted the change on my fork.

Maybe someone else wants to give this a go and see if it's something on my box breaking this PR.

**Comment by ctrl-i** [comment]
> Maybe someone else wants to give this a go and see if it's something on my box breaking this PR.

I could not get this to work on my fork either when I tried a month or so ago.

**Comment by jvolkenant** [comment]
This is quite an addition to MIAB.  I don't expect this PR to be accepted anytime soon.  There may be bugs/improvements to be done to this patch.  I personally need this functionality, so I'll still be using this in my own branch.

**Comment by yodax** [comment]
I hope to come back to this at a later time.

**Comment by m4c3** [comment]
Hi, there are news from Upstream about this: https://salsa.debian.org/java-team/lucene-solr/commit/f58985f500c3ef0c7e874c011ac02ea0c1776de4

**Comment by m4c3** [comment]
i use this for over two weeks now and it works like a charm. is there a way i can help to get this feature in to the official build?

**Comment by jvolkenant** [comment]
I don't foresee that happening.  This is quite a large impact on both performance of low end boxes as well as difficulty added to maintain this project.  I still think there are some improvements to be made.  Some have reported oddities although I have not experience them personally.

**Comment by jvolkenant** [comment]
Should be reviewed and some parts rewritten.  Closing for now.

**Comment by xenithorb** [comment]
I've been using this on a low end digital ocean box with no impact. In fact it's the only way to do meaningful searches with mobile clients. This should be included in the official build 

FTS indexing is expected in modern email search

**Comment by jvolkenant** [comment]
I'm ok with it being an unsupported go outta your way to add if you want it thing.  There are some edge cases that would benefit going over before merging. (Document clearing and reindexing if search stops working, document how to test it, etc.)

I've been using it myself for quite a long time, but I'll admit I don't know that much about jetty/java workings to be able to support it if it really goes wrong.
